### PR TITLE
Add validation for noise channels parameters

### LINKED ISSUE #1337: Noise channels do not validate parameter values, leading to errors
#### Issue description

Description of the issue - include code snippets and screenshots here
if relevant. You may use the following template below

* *Expected behavior:* Raise an error showing that the probability (or sum of probabilities) are not compatible with the expected channel behaviour.

* *Actual behavior:* Raises a `RuntimeWarning` for an invalid value encountered in sqrt.

* *Reproduces how often:* 100%

* *System information:* 
* Name: PennyLane
Version: 0.16.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/nahum/Documents/pennylane-training/pen-env/lib/python3.6/site-packages
Requires: scipy, toml, appdirs, autoray, semantic-version, networkx, autograd, numpy
Required-by: PennyLane-qiskit
Platform info:           Linux-5.4.0-73-generic-x86_64-with-debian-buster-sid
Python version:          3.6.8
Numpy version:           1.19.5
Scipy version:           1.5.4
Installed devices:
- default.gaussian (PennyLane-0.16.0.dev0)
- default.mixed (PennyLane-0.16.0.dev0)
- default.qubit (PennyLane-0.16.0.dev0)
- default.qubit.autograd (PennyLane-0.16.0.dev0)
- default.qubit.jax (PennyLane-0.16.0.dev0)
- default.qubit.tf (PennyLane-0.16.0.dev0)
- default.tensor (PennyLane-0.16.0.dev0)
- default.tensor.tf (PennyLane-0.16.0.dev0)
- qiskit.aer (PennyLane-qiskit-0.15.0)
- qiskit.basicaer (PennyLane-qiskit-0.15.0)
- qiskit.ibmq (PennyLane-qiskit-0.15.0)

#### Source code and tracebacks
```python
from pennylane.ops.channel import AmplitudeDamping

print(AmplitudeDamping(2, wires=0).kraus_matrices)
```

```
pennylane/ops/channel.py:61: RuntimeWarning: invalid value encountered in sqrt
  K0 = np.diag([1, np.sqrt(1 - gamma)])
[array([[ 1.,  0.],
       [ 0., nan]]), array([[0.        , 1.41421356],
       [0.        , 0.        ]])]
```

### PR DESCRIPTION
**Pull request by nahumsa**
**Context:** Noise channels do not validate parameter values

**Description of the Change:** Add validation for noise channel parameters, now it raises an `ValueError` when
  noise parameters are invalid.

**Benefits:** It no longer raises a `RuntimeWarning` when using invalid parameters on noise channels.

**Possible Drawbacks:**

**Related GitHub Issues:** Closes #1337

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 <h3>New features since last release</h3>
 
+* Add validation for noise channel parameters, now it raises an `ValueError` when
+  noise parameters are invalid.
```
Comment (josh146) [inline .github/CHANGELOG.md:6]: Don't forget to add a link back to your PR here!
```suggestion
* Added validation for noise channel parameters. Invalid noise parameters now
   raise a `ValueError`.
  noise parameters are invalid.
```

File: pennylane/ops/channel.py [hunk 2]
```diff
@@ -58,6 +58,10 @@ class AmplitudeDamping(Channel):
     @classmethod
     def _kraus_matrices(cls, *params):
         gamma = params[0]
+
+        if gamma < 0.0 or gamma > 1.0:
```
Comment (josh146) [inline pennylane/ops/channel.py:62]: Alternatively, you could have
```suggestion
        if not 0 <= gamma <= 1.0:
```

I don't think this is any more efficient, but it _might_ be more readable?

**Review by josh146 — COMMENTED** [review]
Thanks for fixing this one @nahumsa 💪 

I've left a couple of preliminary comments, but will bring in @ixfoduap for any extra comments or suggestions.

File: pennylane/ops/channel.py (hunk 2 above)
Comment (nahumsa) [reply pennylane/ops/channel.py:62]: Sure, it looks more readable to me.

File: .github/CHANGELOG.md (hunk 1 above)
Comment (nahumsa) [reply .github/CHANGELOG.md:6]: Yep, forgot about that 😅

**Review by ixfoduap — CHANGES_REQUESTED** [review]
Hi @nahumsa, thanks for another great contribution!

Everything looks in order, but I have an important concern. I remember removing a warning like this because it was throwing errors when the channel parameters where trainable parameters. If I'm not mistaken, the issue was that a check like 
```
if not 0 <= gamma <= 1.0:
```
would fail if `gamma` was not a float. The ability to train noise parameters is a unique feature of PennyLane, as shown in the [noisy circuit tutorial](https://pennylane.ai/qml/demos/tutorial_noisy_circuits.html), so we should be mindful not to break this functionality.

Please add some tests that gradients can be computed correctly and circuits can be trained, and I'll be happy to approve. If the tests fail, then you should still be able to adapt the checks so that no issues occur

**Comment by nahumsa** [comment]
Hi @ixfoduap! 

Thanks for pointing that out! I have a question regarding this test, it would be an end-to-end test just like the example, or just computing the gradient for a given example would be enough?

**Comment by ixfoduap** [comment]
I would like to test that the gradient of a simple circuit is computed correctly. Actually maybe such tests are already done for some of the channels. If that's the case, then I guess no more code needs to be written. Basically I just want to make sure that we don't ship breaking code.

This doesn't have to appear as an actual integration test, but I would be more comfortable if you could run the examples from the tutorial from a pennylane version from this branch and check that everything goes well.

**Comment by nahumsa** [comment]
I think that this test is covered [here](https://github.com/PennyLaneAI/pennylane/blob/12faeca6f7dcd6fcba868e7ae27907d693025476/tests/ops/test_channel_ops.py#L148-L164):

``` python
    @pytest.mark.parametrize("angle", np.linspace(0, 2 * np.pi, 7))
    def test_grad_bitflip(self, angle, tol):
        """Test that analytical gradient is computed correctly for different states. Channel
        grad recipes are independent of channel parameter"""

        dev = qml.device("default.mixed", wires=1)
        prob = 0.5

        @qml.qnode(dev)
        def circuit(p):
            qml.RX(angle, wires=0)
            qml.BitFlip(p, wires=0)
            return qml.expval(qml.PauliZ(0))

        gradient = np.squeeze(qml.grad(circuit)(prob))
        assert gradient == circuit(1) - circuit(0)
        assert np.allclose(gradient, (-2 * np.cos(angle)))
```

Please, let me know if it needs more tests about this.

**Review by ixfoduap — APPROVED** [review]
Thanks for double-checking my concern regarding gradients!

I'm happy to be reminded that we are testing this. Happy to approve from my side

**Review by josh146 — APPROVED** [review]
I tested the gradients locally using both Autograd and TensorFlow, everything continues to work for me 💪 Thanks @nahumsa!
